src/RamTranslator.cpp

```diff
@@ -1059,15 +1086,87 @@ std::unique_ptr<RamStatement> RamTranslator::translateRecursiveRelation(
     }
 
     /* construct fixpoint loop  */
-    return std::unique_ptr<RamStatement>(new RamSequence(std::move(preamble),
-            std::unique_ptr<RamStatement>(new RamLoop(std::move(loopSeq),
-                    std::unique_ptr<RamStatement>(new RamExit(std::move(exitCond))), std::move(updateTable))),
-            std::move(postamble)));
+    std::unique_ptr<RamStatement> res;
+    if (preamble) appendStmt(res, std::move(preamble));
+    if (!loopSeq->getStatements().empty() && exitCond && updateTable) {
+        appendStmt(res, std::unique_ptr<RamStatement>(new RamLoop(std::move(loopSeq),
+                                std::unique_ptr<RamStatement>(new RamExit(std::move(exitCond))),
+                                std::move(updateTable))));
+    } else {
+        if (Global::config().has("fault-tolerance")) {
+            for (const AstRelation* rel : scc) {
+                if (!rel->hasRecordInHead()) {
+                    appendStmt(res, std::unique_ptr<RamStatement>(
+                                            new RamStore(getRamRelationIdentifier(rel, &typeEnv))));
+                }
+            }
+        }
+    }
+    if (postamble) appendStmt(res, std::move(postamble));
+    if (res) return res;
 
     assert(false && "Not Implemented");
     return nullptr;
 }
 
+void createAndLoad(std::unique_ptr<RamStatement>& current, const AstRelation* rel,
+        const TypeEnvironment& typeEnv, const bool isComputed, const bool isRecursive,
+        const bool loadInputOnly) {
+    const auto dir = Global::config().get((rel->isInput() && isComputed) ? "facts-dir" : "output-dir");
+    const auto ext = (!rel->isOutput() || (rel->isInput() && isComputed)) ? ".facts" : ".csv";
+    AstRelation* mrel = const_cast<AstRelation*>(rel)->clone();
+    if ((!rel->isInput() && !loadInputOnly) ||
+            (!isComputed && !isRecursive && !loadInputOnly && rel->isInput())) {
+        std::unique_ptr<AstIODirective> directive = std::unique_ptr<AstIODirective>(new AstIODirective());
+        directive->setAsInput();
+        directive->addKVP("intermediate", "true");
+        mrel->addIODirectives(std::move(directive));
+    }
+    RamRelationIdentifier rrel = getRamRelationIdentifier(
+            mrel, &typeEnv, getRelationName(mrel->getName()), mrel->getArity(), false, dir, ext);
+    // create and load the relation at the start
+    appendStmt(current, std::unique_ptr<RamStatement>(new RamCreate(rrel)));
```

Comment:
make_unique<RamCreate>(rrel) should work, I think, and be simpler. Likewise the unique_ptrs below this.